### DISCUSSION TITLE
TimeSeries: Hide null & and hide zero series

### DISCUSSION THREAD

**Discussion by krocans** [grafana/grafana#38299] [Dashboards, Panels & Visualizations]
Time series panel although it contains visual enhancements lacks what I think important features of old Graph panel:

- hide series with 0/null values
- sort tooltip entries by value.

It would be nice to have those features in the Time series panel especially if it's going to replace Graph panel in future releases.

**Comment by torkelo** [comment]
Question is if this should be implemented as a transform to filter out fields & series or as field override matcher (and then use the hide in option that already exists)

**Reply by amarant** [reply]
This should be a field override matcher the data are already filtered in my case, also the default is very strange, when there is no data why show the graph at 0 ? The old graph used to only show the non null data. It is now not pleasing to show forecast with a special style because all the lines are visible.

![image](https://user-images.githubusercontent.com/2281471/148302961-9070b269-d1f1-4de6-989b-2c287caca9ab.png)

**Comment by pkolyvas** [comment]
We're in the process of moving Grafana feature requests to [GitHub Discussions](https://community.grafana.com/t/were-moving-grafana-feature-requests-to-github-discussions/50341/3). This issue is being moved to [Discussions](https://github.com/grafana/grafana/discussions) or has been moved from our issues list, depending on where you're seeing this message. 

 We hope to use the discussion thread to continue building use cases, and additional evidence based on engagement/voting, in order to help us prioritize this feature request later on down the road. 

As with all things Grafana, we value your input and participation in this project and appreciate your patience during this transitional period. Looking forward to the discussion!

**Comment by faxm0dem** [comment]
I tried in 8.3.3 to achieve this : hide series with only zeroes, but failed.
This is something we make heavy use of, for instance to show only network interfaces with traffic (hiding the unplumbed ones).
I'd really like to be able to do this, or else we have to stick to the graph panel.

**Reply by michaelkebe** [reply]
Looks like a bug.

https://github.com/grafana/grafana/pull/43622

Will be fixed in 3.3.4

**Reply by faxm0dem** [reply]
I don't think this is related. What we want here is this:

![2022-01-18-102234_210x142_scrot](https://user-images.githubusercontent.com/112638/149961711-cf83370b-a581-4bcf-80ff-5fba45421a94.png)

**Comment by nathankopp** [comment]
Please bring back the old functionality.  The LEGEND needs to re-instate the easy-to-use features:

- hide series with only nulls
- hide series with only zeros

The lack of these features is COMPOUNDED by the fact that sorting by columns in a TABLE LEGEND with the highest on top puts the null-valued series ABOVE the numeric series, which is completely USELESS.

**Comment by btlk** [comment]
This feature is a lifesaver and not including it in the "new and more performant" panel is a big oversight.

I do not see myself migrating nice and tidy graphs otherwise

**Comment by meolic** [comment]
Is this already solved? Which version of Grafana and how to use "Hide series with only nulls"?

**Comment by qupterra** [comment]
Has there been any progress on this? 
I am still unable to find an elegant solution for my need to not display series' that are all zeroes or null on this panel within the currently selected time window.

**Comment by Bilanda** [comment]
Hi, is there any news on this ? 
Tried to solve this by upgrading on Grafana 9, but it doesn't fix the issue, this option is still unavailable on new graphs.
Moreover, with Grafana 9, **old graphs are not available anymore** so don't upgrade if you absolutely need these graphs (that's my case :facepalm: )

**Comment by prwood** [comment]
I'm also interested in a solution to this - it's a sorely missing feature in the new Time Series panel.

**Comment by vincenzomanzoni** [comment]
We are also interested to have this feature in Grafana 9. It is very important when you use together with alarms, to show just the timeseries that generate the alarm.

**Comment by tubemeister** [comment]
Would like this too, the new graphs are messy

**Comment by Jojoooo1** [comment]
Any news ?

**Comment by AnnaWojciechowska** [comment]
I also need this feature, the only fix is copying old graph, the 0 zeries are very misleading

**Comment by madiaz-tuenti** [comment]
Any news?

**Comment by CGMBenjamin** [comment]
Is there a workaround to simulate this with a Time Series Panel?

**Comment by rooty0** [comment]
I'm in the same boat. Thinking about downgrade

**Comment by leeoniya** [comment]
this got prioritized today; hold tight. probably will exist as a new Transformer in 9.3

**Reply by leeoniya** [reply]
https://github.com/grafana/grafana/pull/57657

**Comment by cornfeedhobo** [comment]
![image](https://user-images.githubusercontent.com/716304/198885814-3bec65e7-8868-45a6-b886-9e7a4d4ace1c.png)
This sort of worked for me, but maybe someone has a better suggestion.

**Reply by ReLium** [reply]
@cornfeedhobo Where did you find this option? :-)

**Reply by faxm0dem** [reply]
I suppose @cornfeedhobo was referring to the axis limits in the graph options

**Comment by Nadine-Schneider** [comment]
Any news guys? I would like to hide  legends and tooltips for null series that eventually return from a query.

For example, in this case, only two series returned non-null from the filter used in the query. The series that returned null do not appear on the chart, however they appear in the legend and tooltip.

![image](https://user-images.githubusercontent.com/38010089/218841674-8e536ac4-3ebd-4b96-b8e1-a34a8f15556e.png)

Data is retrieved from an SQL query.
And each column returned from the select is a legend in this case. What I want is that the null columns do not appear in the legend or in the tooltip.

Is there already a solution for this?

**Comment by BenoitPoulet** [comment]
They just closed the merge request that was done to fix this 😥
 Filter field by values transformer #57657 

I took 4 month to decline the merge request... and now they have a "better idea" :|
https://github.com/grafana/grafana/pull/64477

As a reminder the feature was removed in **2021**

**Reply by rooty0** [reply]
omg :(

**Comment by leeoniya** [comment]
we just merged https://github.com/grafana/grafana/pull/64477. you should be able to grab a nightly build [1] tomorrow if you'd like to try it out early. please open issues for any bugs.

we have a todo [migration issue](https://github.com/grafana/grafana/issues/64642) to make sure previous settings move over to TimeSeries when we deprecate Graph (old).

there is another related feature request that we will tackle separately: https://github.com/grafana/grafana/issues/6946

[1] https://grafana.com/grafana/download/nightly?edition=oss

**Reply by ImoutoChan** [reply]
When 9.5.0 is expected to be shipped?

**Reply by mrmartan** [reply]
I still have no idea how to use this. This used to be super easy, besides essential.

**Comment by FranckUltima** [comment]
hi all, still no fix for this issue ? i need also to hide null zero series in my legend thanks.

**Comment by Otterpohl** [comment]
tack `!= 0` onto the end of your query if you cant wait for this feature to land 👁️ 👃 👁️

**Reply by Otterpohl** [reply]
@FranckUltima ^^

**Comment by andrgras** [comment]
Grafana 9.5.1, override works well to hide zero series

![image](https://user-images.githubusercontent.com/77773783/236830633-372a3050-9ed8-438f-8980-7bb4bd9ee69f.png)

**Reply by prwood** [reply]
It looked promising, but I'm still seeing a lot of all-zero-value metrics in my graphs. I'm wondering if this is a snag in my implementation: all of the values I'm looking to filter out are non-zero as stored in the graphite database, but in my graphite query I'm applying a nonNegativeDerivative function, so that they end up with a value of zero if they showed no change over the selected period. I'm wondering if this "All Zeros" filter is looking at the base value before the graphite functions are applied. If so, in my case, they'll never look like zero. It *has* effectively filtered out metrics that stopped recording a value at some point in the past and are no longer being updated, but it's not filtering out the metrics that still exist with a non-zero, unchanging value which is calculated as zero by nonNegativeDerivative.

Any thoughts?

ETA: Previous Grafana implementations with the "hide nulls/hide zeros" option had no trouble hiding these same type of metrics that were non-zero before applying nonNegativeDerivative, and zero afterward.

**Reply by prwood** [reply]
Sorry, my mistake - I didn't see that there was also a "All Nulls" option, and in fact the metrics that were being hidden *were* all zeros, and the additional ones I wanted to be hidden were all nulls. After selecting the "All Nulls" option as well, everything is being hidden as desired.

That being said, it did take quite a few more clicks to obtain this functionality than the two clicks required with the older method, but I can see the value of the flexibility inherent in the new method.

**Comment by dixanms** [comment]
In Grafana v9.5.2, the State Timeline visualization does offer to override field with "All Zeros", but the available action don't include a "Hide" or "Remove". I think this should be implemented as a Transformation we can easily apply to any visualizations.

**Comment by gelicia** [comment]
Hello! We want to let you all know that we are moving to Github Issues for feature requests. You can read more about the move [here](https://github.com/grafana/grafana/discussions/73424).

This feature seems to be partially resolved, so I am going to close this broader discussion. If there is an aspect of this feature that was not addressed, please create a new issue detailing what you'd like to see done, and it will be triaged and prioritized along with the team's other work. 

If you'd like to see the work that was already merged addressing this, please reference 
Conditionally hide 0s and nulls - https://github.com/grafana/grafana/pull/64477 
Migrate hiding all nulls or all 0s - https://github.com/grafana/grafana/issues/64642
